### PR TITLE
Refactor event log to be Vector instead of Seq (List)

### LINKED ISSUE #3287: Refactor event log to be Vector instead of Seq (List)
## Overview
We should investigate if using `Vector` for event log will speed up overall execution of Rholang when creating and replaying a block.

https://github.com/rchain/rchain/blob/9ef35159ee0ca5291b8ffdf769c9522a05ab6d9f/rspace/src/main/scala/coop/rchain/rspace/trace/package.scala#L7

### PR DESCRIPTION
**Pull request by vinay-bhamidipati**
## Overview

Addresses  #3287. Makes event log a Vector rather than a Seq. Fixes related issues with this change. Targeted toward improving performance.

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by vinay-bhamidipati** [comment]
Further improvements may be made by removing toVector in some places where earlier toList was being used to convert a Seq to a List. From what I can see, this would have no conflicts, but I held off from doing it since I was unsure whether it would be necessary for other parts of the project. I can remove them if permissible, let me know.

**Comment by vinay-bhamidipati** [comment]
Also, any input on how to actually determine whether these changes increase perf?

**Comment by tgrospic** [comment]
bors try

File: .github/workflows/continuous-integration.yml
Comment (tgrospic) [inline .github/workflows/continuous-integration.yml]: Previously we had different dependency for `try` and `merge` commands. For the _try_ only tests were run and for _merge_ also Docker image is published as mentioned in comments.

Can we have the same functionality as before or it's how bors works now?

Also I'd suggest that this change is done in a new PR and after merge this PR can be rebased.

File: project/Dependencies.scala
Comment (tgrospic) [inline project/Dependencies.scala]: Changes to dependency should not be in this PR. Also please squash/rebase commits so that intermediate changes will not be merged.

File: models/src/main/scala/coop/rchain/models/HashM.scala
Comment (tgrospic) [inline models/src/main/scala/coop/rchain/models/HashM.scala]: `Vector` and `List` extends `Seq` so the new definition specific to Vector should not be needed. What is the reason to add these implementations for equals and hash code?

File: .github/workflows/continuous-integration.yml
Comment (vinay-bhamidipati) [reply .github/workflows/continuous-integration.yml]: With bors now, there can be only one test named 'bors build finished'. If there are multiple and one or more skips, bors considers it a failure. One way to fix it would be to have 'release docker image' execute as a standalone test for ref == staging and have bors_success just depend on unit and integration tests. That way, if release docker image fails on a merge, bors will fail, but on a try, 'release docker image' wont execute and bors will succeed. Hopefully.

File: models/src/main/scala/coop/rchain/models/HashM.scala
Comment (vinay-bhamidipati) [reply models/src/main/scala/coop/rchain/models/HashM.scala]: Without them, the code will not compile; several errors will appear in those files. I think it is because in the definition in HashM and EqualsM for Seq the type is explicitly seq and does not use < to indicate that child classes can be used. I don't think Lists were being used by EqualsM and HashM, I think it was Seq that was being used directly so there were no conflicts with the old code.

Comment (tgrospic) [reply models/src/main/scala/coop/rchain/models/HashM.scala]: Makes sense. This will also ensure that no conversion is happening here. `value` is already and Vector so `.toVector` is redundant.

For comprehension is not needed, at the end it's just mapping `.map(_.hashCode)`.

File: casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala
Comment (tgrospic) [inline casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala]: `systemDeploys` will be a List here so better to not change it. Although this list should not be big enough to make a difference.

P.S.
With latest cats update it's no longer needed to import instance when `traverse` (`foldM`) is used. In cats v3 (or 2.3) Seq will have instance for Traverse so explicit conversion should also be redundant.
https://github.com/typelevel/cats/pull/3620

File: models/src/main/scala/coop/rchain/casper/protocol/CasperMessage.scala
Comment (tgrospic) [inline models/src/main/scala/coop/rchain/casper/protocol/CasperMessage.scala:346]: Is `deployLog` already a Vector?

File: casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala
Comment (vinay-bhamidipati) [reply casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala]: Changed!

File: models/src/main/scala/coop/rchain/models/HashM.scala
Comment (vinay-bhamidipati) [reply models/src/main/scala/coop/rchain/models/HashM.scala]: Changed!

File: models/src/main/scala/coop/rchain/casper/protocol/CasperMessage.scala
Comment (vinay-bhamidipati) [reply models/src/main/scala/coop/rchain/casper/protocol/CasperMessage.scala:346]: deployLog on the right comes from the class ProcessedDeployProto which defines it to be a Seq. That is why toVector is necessary so that it is converted from Seq to Vector. Maybe we make deployLog in ProcessedDeployProto a vector instead of Seq as well? @tgrospic

**Comment by nzpr** [comment]
@vb-cs to see influence on perf, if you don't want manually measure parts of the code you are tuning, you should use jaeger. The simplest way to do that is to add `--zipkin` to your run command, and run jaeger instance locally https://www.jaegertracing.io/docs/1.21/getting-started/#all-in-one - its just one command and you have instance running locally. 

Rnode with `--zipkin` flag streams events to localhost jaeger server, so you will see events there. 
It will look like this, list of events:
Screenshot 2021-01-08 at 11 14 36
and detailed view of one particular event of block replay
Screenshot 2021-01-08 at 11 15 02

So you can see what is the difference there. 
You will need to create blocks on your local node, so I'm assuming you know how to do that.

**Comment by 9rb** [comment]
bors try

File: .github/workflows/continuous-integration.yml
Comment (tgrospic) [reply .github/workflows/continuous-integration.yml]: I think we can release Docker image on bors Try also.

**Comment by tgrospic** [comment]
@vb-cs everything now looks good. Please add some data as confirmation that these changes resulted in performance speedup. Measurement can be done with Jaeger as mentioned by @nzpr. Logs data can also be used to show the difference in duration of some operations e.g. genesis creation.

**Comment by vinay-bhamidipati** [comment]
> 
> 
> @vb-cs everything now looks good. Please add some data as confirmation that these changes resulted in performance speedup. Measurement can be done with Jaeger as mentioned by @nzpr. Logs data can also be used to show the difference in duration of some operations e.g. genesis creation.

Having trouble with this. When I try to run rnode I get 

node 14:19:35.626 [ERROR] [main                ] [coop.rchain.node.Main$       ] - Unhandled exception in thread main
node coop.rchain.blockstorage.finality.LastFinalizedBlockNotFound$: null

@tgrospic

**Review by 9rb — APPROVED** [review]
If @zsluedem or @nzpr can confirm the performance benefits, this can be merged after a second approval

**Comment by zsluedem** [comment]
I don't see big difference in performance in my tests

**Review by zsluedem — APPROVED** [review]
LGTM

**Comment by tgrospic** [comment]
If there is no difference in performance better to leave it as is because `Seq` is more general then `Vector`.

**Comment by zsluedem** [comment]
> If there is no difference in performance better to leave it as is because `Seq` is more general then `Vector`.

@tgrospic  I am curious in what way you tested see performance better.

**Comment by 9rb** [comment]
> 
> 
> I don't see big difference in performance in my tests

@zsluedem could you please post some performance test data here, pre- and post- this PR ? 
In Scala 3.0 there's ArraySeq and other data structures. I think @tgrospic may be implying that we can try those later if vector doesn't increase performance and then make a choice of which structure to use.

**Comment by zsluedem** [comment]
Get block and replay a block avg time:
Vector:  10.040824460983277 s
Seq: 10.043872594833374  s

**Comment by 9rb** [comment]
Closing this for now, since the performance improvement seems miniscule. We can revisit this issue with ArraySeq and other structures in Scala 2.13 and Scala 3.0 and re-open this at that time, if needed

**Pull request closed without merging**
